### ISSUE TITLE
percpu accessors have load hit store issues

### ISSUE THREAD

**Issue by antonblanchard** [linuxppc/linux#45]
I traced some load hit store issues back to the percpu data accessors. An example:

```
unsigned long foo(unsigned long __percpu *p)
{
        return __this_cpu_inc_return(*p);
}
```

produces:

```
0000000000002690 <foo>:
    2690:       30 00 4d e9     ld      r10,48(r13)
    2694:       2a 50 23 7d     ldx     r9,r3,r10
    2698:       01 00 29 39     addi    r9,r9,1
    269c:       2a 51 23 7d     stdx    r9,r3,r10      <--- store
    26a0:       30 00 2d e9     ld      r9,48(r13)
    26a4:       2a 18 69 7c     ldx     r3,r9,r3         <--- load
    26a8:       20 00 80 4e     blr
```

This is because the generic version of __this_cpu_inc_return is an add followed by a read:

```
#define raw_cpu_generic_add_return(pcp, val)                            \
({                                                                      \
        raw_cpu_add(pcp, val);                                          \
        raw_cpu_read(pcp);                                              \
})
```

The fact we load the percpu data pointer twice is a little disappointing too.

**Comment by npiggin** [comment]
On powerpc, the code basically turns into:
```
*(&pcp + local_paca->data_offset/sizeof(pcp)) += val;
return *(&pcp + local_paca->data_offset/sizeof(pcp));
```

The reason for the poor  code is that the compiler assumes the stores through the pointer may alias data_offset, so that has to be reloaded. It also assumes local_paca can be aliased, even though it's a register variable. Restrict pointers can be used to solve this, but they're tricky to do (can't cast to a restrict pointer). Rewriting the generic RMW functions to calculate the pointer address just once is probably the best short term fix.

```
        ld 8,48(13)      # _4->data_offset, _4->data_offset
        mr 9,3   # __ptr, p
        ldx 10,9,8       # *_7, *_7
        addi 10,10,1     # D.44882, *_7,
        mr 3,10  #, D.44882
        stdx 10,9,8      # *_7, D.44882
        blr
```

Using a register is even better (in this case, r14)

```
        ldx 9,3,14       # *_4, *_4
        mr 8,3   # p, p
        addi 9,9,1       # D.45041, *_4,
        mr 3,9   #, D.45041
        stdx 9,8,14      # *_4, D.45041
        blr
```

This is were we ideally want to be for percpu data.

We could possibly even reclaim r13 for this usage, by keeping a pointer to paca in a per-cpu area. Performance critical fields from the paca, such as paca_index could be moved or duplicated as per-cpu data to avoid an additional load. This would require going through a vast amount of assembly, however.

**Comment by antonblanchard** [comment]
A stupid little program to look at our register usage in the kernel:
```
#!/usr/bin/python

import sys
import subprocess
import re
import operator


vmlinux = sys.argv[1]

cmd = ['objdump', '-d', vmlinux ]
pipe = subprocess.Popen(cmd, stdout=subprocess.PIPE, close_fds=True).stdout

r1 = re.compile('(r[0-9]*)')
r2 = re.compile('[-0-9]*\((r[0-9]*)\)')

registers = {}
for i in range(32):
        registers["r%d" % i] = 0


for line in pipe.readlines():
        x = line.split()
        if len(x) < 7:
                continue

        x = x[6].split(',')

        for entry in x:
                m = r1.match(entry)
                if m:
                        register = m.group(1)
                        registers[register] = registers[register] + 1
                        continue

                m = r2.match(entry)
                if m:
                        register = m.group(1)
                        registers[register] = registers[register] + 1
                        continue


for (register, count) in sorted(registers.items(), key=operator.itemgetter(1), reverse=True):
        print "%s\t%d" % (register, count)
```

**Comment by antonblanchard** [comment]
As expected, r14 is hardly used (it's the last non volatile GPR to be used by gcc). Perhaps we can just take it over via -ffixed-r14, which allows us to move things out of the paca incrementally.

```
r9	576776
r1	501029
r3	327822
r2	308430
r31	267739
r10	262070
r30	176590
r4	157157
r0	138273
r8	123137
r29	120772
r5	103353
r28	84694
r7	72162
r6	64412
r27	60969
r12	47728
r26	45466
r25	33738
r24	26608
r23	21308
r13	21051
r22	16734
r21	13654
r11	12431
r20	10649
r19	8940
r18	7859
r16	7178
r17	7009
r15	6397
r14	6057
```

**Comment by mpe** [comment]
![register-usage](https://cloud.githubusercontent.com/assets/54119/15955475/dae654ae-2f24-11e6-97e5-006f57551439.png)

**Comment by npiggin** [comment]
This is fixed by commit 1b5ca12127427c51be605a75ecd0141eb3357249 ("percpu: improve generic percpu modify-return implementation")

**Comment by antonblanchard** [comment]
Seeing this has been fixed for over two years, we may as well close it.
